auslib/db.py

```diff
@@ -854,80 +896,98 @@ def insert(self, changed_by, transaction=None, dryrun=False, **columns):
+                kwargs["columns"].append(self.sc_id)
+        for row in super(ScheduledChangeTable, self).select(where=where, transaction=transaction, **kwargs):
+            columns = [getattr(self.conditions, c) for c in itertools.chain(*self.conditions.condition_groups)]
+            conditions = self.conditions.select([self.conditions.sc_id == row["sc_id"]], transaction=transaction, columns=columns)
+            row.update(conditions[0])
+            ret.append(row)
+        return ret
 
-        # Use the appropriate base table methods in dry run mode to ensure the
-        # user has permission for the change they want to schedule
-        if columns.get("data_version"):
-            self.baseTable.update(base_table_where, columns, changed_by, columns["data_version"], transaction=transaction, dryrun=True)
-        else:
-            self.baseTable.insert(changed_by, transaction=transaction, dryrun=True, **columns)
+    def insert(self, changed_by, transaction=None, dryrun=False, **columns):
+        base_columns, condition_columns = self._splitColumns(columns)
+
+        self.validate(base_columns, condition_columns, changed_by, transaction)
 
+        base_columns = self._prefixColumns(base_columns)
+        base_columns["scheduled_by"] = changed_by
+        ret = super(ScheduledChangeTable, self).insert(changed_by=changed_by, transaction=transaction, dryrun=dryrun, **base_columns)
         if not dryrun:
-            what["scheduled_by"] = changed_by
-            ret = super(ScheduledChangeTable, self).insert(changed_by=changed_by, transaction=transaction, **what)
-            return ret.inserted_primary_key[0]
+            sc_id = ret.inserted_primary_key[0]
+            self.conditions.insert(changed_by, transaction, dryrun, sc_id=sc_id, **condition_columns)
+            return sc_id
 
     def update(self, where, what, changed_by, old_data_version, transaction=None, dryrun=False):
+        base_what, condition_what = self._splitColumns(what)
+
         # We need to check each Scheduled Change that would be affected by this
         # to ensure the new row will be valid.
         for row in self.select(where=where, transaction=transaction):
-            new_row = row.copy()
-            new_row.update(self._prefixColumns(what))
-            base_table_where = {pk: new_row["base_%s" % pk] for pk in self.base_primary_key}
-            if new_row.get("base_data_version"):
-                self.baseTable.update(base_table_where, new_row, changed_by, new_row["base_data_version"], transaction=transaction, dryrun=True)
-            else:
-                self.baseTable.insert(changed_by, transaction=transaction, dryrun=True, **new_row)
+            # Before validation, we need to create the new version of the
+            # Scheduled Change by combining the old one with the new data.
+            sc_columns, condition_columns = self._splitColumns(row)
+            base_columns = {}
```

Comment:
Hm, I did this, but then realized it might be better to combine the two existing loops into one - since the purpose of both is to get the new base_columns. The first loop (or dict comprehension, if we use that) is really only an intermediate state. What do you think of:
```
            base_columns = {}
            for col in sc_columns:
                if not col.startswith("base_"):
                    continue
                base_col = col.replace("base_", "")
                if base_col in base_what:
                    base_columns[base_col] = base_what[base_col]
                elif sc_columns.get(col):
                    base_columns[base_col] = sc_columns[col]
```